### PR TITLE
Stabilize `Rc`, `Arc` and `Pin` as method receivers

### LINKED ISSUE #55786: [Stabilization] Stablize using some arbitrary self types defined in std
Feature name: `arbitrary_self_types`
Stabilization target: 1.32.0
Tracking issue: #44874
Related RFCs: rust-lang/rfcs#2362

This is a proposal to stabilize a subset of the `arbitrary_self_types` feature,
making it possible on stable to use valid "arbitrary" self types that have been
defined in the standard library, but not to define your own new self types.

## Stabilized feature or APIs

Today, only the following types are allowed as self types unless the user uses
the `arbitrary_self_types` flag:

* `Self`
* `&Self`
* `&mut Self`
* `Box<Self>`

We've long desired to extend this set to all pointer types defined in std, and
(ideally) to arbitrary user defined pointer types. For quite a while now, the
ability to create user defined self types has existed on nightly under the
`arbitrary_self_types` flag; in this stabilization, we propose to stabilize the
extension of self types to all the relevant pointer types defined in std ~and
their compositions~, while leaving the ability to create your own self types
unstable for now while we iterate on the exact requirements.

The new self types that will be enabled by this stabilization are:

* ~`*const Self`~
* ~`*mut Self`~
* `Rc<Self>`
* `Arc<Self>`
* `Pin<P>` where `P` is another type in this set.
* ~The composition of any members in this set (e.g. `&Box<Self>`,
  `Pin<&mut Rc<Self>>`).~

Additionally, all of these receiver types except for `Self` are object safe, in
the sense that they can be used as the receivers of object-safe trait methods.

### ~Object safety of raw pointer types~

~By making pointers object-safe, we have introduced an additional requirement on
raw pointers: a wide raw pointer must contain valid metadata, even if the
data pointer is null or dangling. That is, a `*const dyn Trait` must have
a metadata pointer pointing to a valid vtable for `Trait`.~

~As an alternative, we could possibly restrict this to only allowing raw
pointers as the receiver types for unsafe methods, and validity of the metadata
pointer would be an invariant the caller would be expected to uphold.~

### Non-`Deref` pointer types

Library defined pointer types can only be receiver types if they implement
`Deref` currently. This excludes certain std pointer types that don't implement
`Deref`, because they could be dangling:

* `NonNull<Self>`
* `rc::Weak<Self>`
* `sync::Weak<Self>`

This stabilization is forward compatible with someday supporting these pointers
as receiver types as we continue to iterate on the requirements of defining
arbitrary self types.

### Magic traits involved

It's worth noting that currently the implementation of this functionality
involves the intersection of several ops traits:

* `Deref` - a self type must be deref, transitively targeting `Self`
* `CoerceUnsized`
* `DispatchFromDyn` - this and the previous are necessary for object safe
* (potentially) `Receiver` - a private trait in std used to limit stabilization
  only to these std traits

The interaction of this feature with these traits is *not* made stable as a
part of this proposal. The traits listed that are still unstable remain
unstable. We still have flexibility to iterate on the exact requirements on
arbitrary self types so long as those requirements include all of the types I
have enumerated previously.

## Implementation changes prior to stabilization

- [ ] `Receiver` trait to limit stable self types to those defined in std
  (@mikeyhew is working on this)

- [ ] Adjust documentation and diagnostics to note the additional valid
  receiver types defined in std, instead of only the 4 accepted today

## Connected features and larger milestones

A trait making use of this feature is the `Future` trait, which is still
unstable under the `futures_api` feature:

```rust
trait Future {
    type Output;
    fn poll(self: Pin<&mut Self>, lw: &LocalWaker) -> Poll<Self::Output>;
}
```

Before we can stabilize this trait, we need to stabilize the ability to
implement traits using `Pin<&mut Self>` as a receiver type: hence, this
stabilization. Stabilizing the `futures_api` feature is a high priority because
it is a blocker for stabilizing the async/await syntax, which makes nonblocking
IO much more ergonomic for Rust.

cc @mikeyhew

### PR DESCRIPTION
**Pull request by mikeyhew**
Replaces #55880
Closes  #55786 
r? @nikomatsakis 
cc @withoutboats @cramertj

This lets you write methods using `self: Rc<Self>`, `self: Arc<Self>`, `self: Pin<&mut Self>`, `self: Pin<Box<Self>`, and other combinations involving `Pin` and another stdlib receiver type, without needing the `arbitrary_self_types`. Other user-created receiver types can be used, but they still require the feature flag to use.

This is implemented by introducing a new trait, `Receiver`, which the method receiver's type must implement if the `arbitrary_self_types` feature is not enabled. To keep composed receiver types such as `&Arc<Self>` unstable, the receiver type is also required to implement `Deref<Target=Self>` when the feature flag is not enabled.

This lets you use `self: Rc<Self>` and `self: Arc<Self>` in stable Rust, which was not allowed previously. It was agreed that they would be stabilized in #55786. `self: Pin<&Self>` and other pinned receiver types do not require the `arbitrary_self_types` feature, but they cannot be used on stable because `Pin` still requires the `pin` feature.

### CODE REVIEW

File: src/librustc_typeck/check/wfcheck.rs [hunk 1]
```diff
@@ -749,72 +748,172 @@ fn check_method_receiver<'fcx, 'gcx, 'tcx>(fcx: &FnCtxt<'fcx, 'gcx, 'tcx>,
...
+            receiver_trait_ref.to_predicate()
+        );
+
+        if !fcx.predicate_must_hold(&receiver_obligation) {
+            debug!("receiver_is_valid: type `{:?}` does not implement `Receiver` trait",
+                receiver_ty);
+            return false
+        }
+
+        let deref_trait_def_id = match fcx.tcx.lang_items().deref_trait() {
```
Comment (nikomatsakis) [inline src/librustc_typeck/check/wfcheck.rs]: It'd be nice to have some comments here, to document what we are requiring and *why* (basically what's in your PR comment).

Comment (nikomatsakis) [reply src/librustc_typeck/check/wfcheck.rs]: This can reference the doc comment on the fn itself I guess.

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
Looks good to me. One nit, then r=me.

**Comment by bluss** [comment]
Since Pin is unstable, does this keep `fn foo(self: Pin<Box<Self>>)` or similar as unstable to use?

**Comment by mikeyhew** [comment]
@nikomatsakis assuming this passes CI, it should be good to go

**Comment by mikeyhew** [comment]
This is good to go again.

**Comment by alexcrichton** [comment]
@bors: retry

* https://github.com/malept/crypto-hash/issues/6

**Comment by cramertj** [comment]
@bors retry

```
[02:58:21] [ 46%] Building CXX object lib/IR/CMakeFiles/LLVMCore.dir/Instruction.cpp.o
[02:58:32] [ 46%] Building CXX object lib/CodeGen/CMakeFiles/LLVMCodeGen.dir/DeadMachineInstructionElim.cpp.o
[02:58:42] [ 46%] Building CXX object lib/CodeGen/AsmPrinter/CMakeFiles/LLVMAsmPrinter.dir/DebugLocStream.cpp.o

Broadcast message from root@travis-job-3fdf3aa9-85be-4ce0-8bb9-67e639ddc737
 (unknown) at 9:25 ...
The system is going down for power off NOW!
```
:(

**Pull request merged**